### PR TITLE
Changed pointers to references where `nullptr` is invalid

### PR DESCRIPTION
**Pull request by mitchellstern**
This standardizes the code base according to @neubig's suggestion in #201. A few minor warnings are also fixed.

### CODE REVIEW

**Comment by neubig** [comment]
Thanks! This looks good. This will break external software toolkits that rely on DyNet's C++ API, so I think we should notify people on the dynet-users mailing list, and then pull in a day or so.

**Comment by mitchellstern** [comment]
Sounds good, thanks! I think the major API change to note is that all builders (e.g. all RNN and softmax variants) and all trainers now take a `Model&` instead of a `Model*` during construction. In addition, a few utility functions (e.g. in `dict.h` and `pretrain.h`) were modified to take references instead of pointers where appropriate.
